node/core/pvf/src/artifacts.rs

```diff
@@ -0,0 +1,288 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::LOG_TARGET;
+use always_assert::always;
+use async_std::{
+	io,
+	path::{Path, PathBuf},
+};
+use polkadot_core_primitives::Hash;
+use std::{
+	collections::HashMap,
+	time::{Duration, SystemTime},
+};
+use parity_scale_codec::{Encode, Decode};
+use futures::StreamExt;
+
+/// A final product of preparation process. Contains either a ready to run compiled artifact or
+/// a description what went wrong.
+#[derive(Encode, Decode)]
+pub enum Artifact {
+	/// During the prevalidation stage of preparation an issue was found with the PVF.
+	PrevalidationErr(String),
+	/// Compilation failed for the given PVF.
+	PreparationErr(String),
+	/// This state indicates that the process assigned to prepare the artifact wasn't responsible
+	/// or were killed. This state is reported by the validation host (not by the worker).
+	DidntMakeIt,
+	/// The PVF passed all the checks and is ready for execution.
+	Compiled {
+		compiled_artifact: Vec<u8>,
+	},
+}
+
+impl Artifact {
+	/// Serializes this struct into a byte buffer.
+	pub fn serialize(&self) -> Vec<u8> {
+		self.encode()
+	}
+
+	/// Deserialize the given byte buffer to an artifact.
+	pub fn deserialize(mut bytes: &[u8]) -> Result<Self, String> {
+		Artifact::decode(&mut bytes).map_err(|e| format!("{:?}", e))
+	}
+}
+
+/// NOTE if we get to multiple engine implementations the artifact ID should include the engine
+/// type as well.
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct ArtifactId {
+	code_hash: Hash,
+}
+
+impl ArtifactId {
+	const PREFIX: &'static str = "wasmtime_1_";
+
+	pub fn new(code_hash: Hash) -> Self {
+		Self { code_hash }
+	}
+
+	/// Tries to recover the artifact id from the given file name.
+	pub fn from_file_name(file_name: &str) -> Option<Self> {
+		use std::str::FromStr as _;
+
+		let file_name = file_name.strip_prefix(Self::PREFIX)?;
+		let code_hash = Hash::from_str(file_name).ok()?;
+
+		Some(Self { code_hash })
+	}
+
+	/// Returns the expected path to this artifact given the root of the cache.
+	pub fn path(&self, cache_path: &Path) -> PathBuf {
+		let file_name = format!("{}{}", Self::PREFIX, self.code_hash.to_string());
+		cache_path.join(file_name)
+	}
+}
+
+#[test]
+fn ensure_wasmtime_version() {
```

Comment:
I understand that this is a somewhat special test, but why not move it to the tests module?

Reply:
Const panics would be nice for this.

Reply:
I approached this with `static_assertions` first and sadly discovered that it only works with numbers. You cannot unfortunately make string comparisons in the const context.

Yeah, this is to make it closer to the `ArtifactId`, don't mind moving it downstairs under the tests module though. Just ask

Reply:
Yeah, would be cleaner IMHO.